### PR TITLE
Tailsitter attitude correction

### PR DESCRIPTION
**Pull request by amandastowers**
Hi, 
I have added some logic to do orientation conversion on attitude for tailsitters in fixed wing mode. Before this logic, tailsitters would appear very unable to track attitude trajectories in fixed wing mode. Now flight review can correctly display the tailsitter attitudes. See attached images for before and after.
newpitch
newroll
oldpitch
oldroll

### CODE REVIEW

**Comment by bkueng** [comment]
Thanks for looking into this. I'm a bit surprised to see the FC does not publish a consistent attitude + setpoint. Instead of adding custom logic in here, we should make sure the data is published consistently.
I had a quick chat with @sfuhrer and he anyway wants to do that.

**Review by sfuhrer — COMMENTED** [review]
It may take a while till we have a consistent attitude setpoint publishing on PX4, so I would bring this in. Looks pretty clean to me.

File: app/plot_app/configured_plots.py [hunk 1]
```diff
@@ -76,6 +78,11 @@ def generate_plots(ulog, px4_ulog, db_data, vehicle_data, link_to_3d_page,
         cur_dataset = ulog.get_dataset('vehicle_status')
         if np.amax(cur_dataset.data['is_vtol']) == 1:
             is_vtol = True
+            # check if is tailsitter
+            if np.amax(cur_dataset.data['is_vtol_tailsitter']) ==1:
```
Comment (bkueng) [inline app/plot_app/configured_plots.py:82]:
```suggestion
            is_vtol_tailsitter = np.amax(cur_dataset.data['is_vtol_tailsitter']) ==1
```

File: app/plot_app/configured_plots.py [hunk 2]
```diff
@@ -76,6 +78,11 @@ def generate_plots(ulog, px4_ulog, db_data, vehicle_data, link_to_3d_page,
         cur_dataset = ulog.get_dataset('vehicle_status')
```
Comment (bkueng) [inline app/plot_app/configured_plots.py:78]: Above, add `is_vtol_tailsitter = False` next to `is_vtol = False`

File: app/plot_app/configured_plots.py [hunk 3]
```diff
@@ -186,19 +193,97 @@ def generate_plots(ulog, px4_ulog, db_data, vehicle_data, link_to_3d_page,
 
     if data_plot.finalize() is not None: plots.append(data_plot)
 
-
+    # VTOL tailistter orientation conversion, if relevant
+    if is_vtol_tailsitter:
+        print(vtol_states)
```
Comment (bkueng) [inline app/plot_app/configured_plots.py:198]: Can you remove the print and move this into a separate method in a new file `vtol_tailsitter.py`?

File: app/plot_app/configured_plots.py [hunk 4]
```diff
@@ -186,19 +193,97 @@ def generate_plots(ulog, px4_ulog, db_data, vehicle_data, link_to_3d_page,
...
+        for d in data:
+            if d.name == 'vehicle_attitude':
```
Comment (bkueng) [inline app/plot_app/configured_plots.py:200]: If the log does not contain either of these topics there's undefined variables below. Make sure that is handled gracefully.

File: app/plot_app/configured_plots.py [hunk 5]
```diff
@@ -186,19 +193,97 @@ def generate_plots(ulog, px4_ulog, db_data, vehicle_data, link_to_3d_page,
...
+
+        # temporary variables for storing VTOL states
+        is_FW = False
+        FW_start = np.nan
+        FW_end = np.nan
+
+        for i in vtol_states:
+            # states: 1=transition, 2=FW, 3=MC
+            # if in FW mode then used FW conversions 
+            if is_FW == True:
```
Comment (bkueng) [inline app/plot_app/configured_plots.py:246]:
```suggestion
            if is_FW:
```

**Review by bkueng — COMMENTED** [review]
Alright I'm ok with getting it in and remove it later on once PX4 is updated.

**Comment by amandastowers** [comment]
Changes requested by @bkueng incorporated.

File: app/plot_app/configured_plots.py [hunk 6]
```diff
@@ -220,9 +234,15 @@ def generate_plots(ulog, px4_ulog, db_data, vehicle_data, link_to_3d_page,
                              y_axis_label='[deg/s]', title=axis_name+' Angular Rate',
                              plot_height='small', changed_params=changed_params,
                              x_range=x_range)
-        data_plot.add_graph([lambda data: (axis+'speed',
-                                           np.rad2deg(data[rate_field_names[index]]))],
-                            colors3[0:1], [axis_name+' Rate Estimated'], mark_nan=True)
+        if is_vtol_tailsitter:
+            try:
```
Comment (bkueng) [inline app/plot_app/configured_plots.py:238]: Why is this needed?

File: app/plot_app/vtol_tailsitter.py [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+# tailsitter uses 90 degree rotation in pitch which is hardcoded in rather than consistently reported in estimated and setpoint
+# use setpoint values as ground truth here and correct estimated by 90 degrees
+
+from scipy.spatial.transform import Rotation as Rot
+import numpy as np
+
+def correct_tailsitter_attitude_and_rates(data,vtol_states):
+
+    for d in data:
+        if d.name == 'vehicle_attitude':
```
Comment (bkueng) [inline app/plot_app/vtol_tailsitter.py:11]: Can you add some flags to check if the topics were found?

**Review by bkueng — COMMENTED** [review]
Thanks for updating.
There's a couple of CI style failures:
```
************* Module plot_app.configured_plots
plot_app/configured_plots.py:196:79: C0303: Trailing whitespace (trailing-whitespace)
plot_app/configured_plots.py:208:0: C0301: Line too long (138/100) (line-too-long)
plot_app/configured_plots.py:214:0: C0303: Trailing whitespace (trailing-whitespace)
plot_app/configured_plots.py:215:0: C0303: Trailing whitespace (trailing-whitespace)
plot_app/configured_plots.py:239:0: C0301: Line too long (144/100) (line-too-long)
plot_app/configured_plots.py:196:9: C0103: Variable name "RPY" doesn't conform to '[a-z_][a-z0-9_]{2,30}$' pattern (invalid-name)
************* Module plot_app.vtol_tailsitter
plot_app/vtol_tailsitter.py:2:0: C0301: Line too long (127/100) (line-too-long)
plot_app/vtol_tailsitter.py:56:44: C0303: Trailing whitespace (trailing-whitespace)
plot_app/vtol_tailsitter.py:59:0: C0301: Line too long (104/100) (line-too-long)
plot_app/vtol_tailsitter.py:60:0: C0301: Line too long (106/100) (line-too-long)
plot_app/vtol_tailsitter.py:61:0: C0301: Line too long (102/100) (line-too-long)
plot_app/vtol_tailsitter.py:62:0: C0301: Line too long (106/100) (line-too-long)
plot_app/vtol_tailsitter.py:79:51: C0303: Trailing whitespace (trailing-whitespace)
plot_app/vtol_tailsitter.py:1:0: C0114: Missing module docstring (missing-module-docstring)
plot_app/vtol_tailsitter.py:8:0: C0116: Missing function or method docstring (missing-function-docstring)
plot_app/vtol_tailsitter.py:8:0: C0103: Function name "correct_tailsitter_attitude_and_rates" doesn't conform to '[a-z_][a-z0-9_]{2,30}$' pattern (invalid-name)
plot_app/vtol_tailsitter.py:12:12: C0103: Variable name "q0" doesn't conform to '[a-z_][a-z0-9_]{2,30}$' pattern (invalid-name)
plot_app/vtol_tailsitter.py:13:12: C0103: Variable name "q1" doesn't conform to '[a-z_][a-z0-9_]{2,30}$' pattern (invalid-name)
plot_app/vtol_tailsitter.py:14:12: C0103: Variable name "q2" doesn't conform to '[a-z_][a-z0-9_]{2,30}$' pattern (invalid-name)
plot_app/vtol_tailsitter.py:15:12: C0103: Variable name "q3" doesn't conform to '[a-z_][a-z0-9_]{2,30}$' pattern (invalid-name)
plot_app/vtol_tailsitter.py:16:12: C0103: Variable name "qt" doesn't conform to '[a-z_][a-z0-9_]{2,30}$' pattern (invalid-name)
plot_app/vtol_tailsitter.py:24:4: C0103: Variable name "RPY" doesn't conform to '[a-z_][a-z0-9_]{2,30}$' pattern (invalid-name)
plot_app/vtol_tailsitter.py:26:4: C0103: Variable name "FW_rotation" doesn't conform to '[a-z_][a-z0-9_]{2,30}$' pattern (invalid-name)
plot_app/vtol_tailsitter.py:27:4: C0103: Variable name "RPY_FW" doesn't conform to '[a-z_][a-z0-9_]{2,30}$' pattern (invalid-name)
plot_app/vtol_tailsitter.py:50:4: C0103: Variable name "is_FW" doesn't conform to '[a-z_][a-z0-9_]{2,30}$' pattern (invalid-name)
plot_app/vtol_tailsitter.py:51:4: C0103: Variable name "FW_start" doesn't conform to '[a-z_][a-z0-9_]{2,30}$' pattern (invalid-name)
plot_app/vtol_tailsitter.py:52:4: C0103: Variable name "FW_end" doesn't conform to '[a-z_][a-z0-9_]{2,30}$' pattern (invalid-name)
plot_app/vtol_tailsitter.py:58:12: C0103: Variable name "FW_end" doesn't conform to '[a-z_][a-z0-9_]{2,30}$' pattern (invalid-name)
plot_app/vtol_tailsitter.py:64:12: C0103: Variable name "is_FW" doesn't conform to '[a-z_][a-z0-9_]{2,30}$' pattern (invalid-name)
plot_app/vtol_tailsitter.py:66:12: C0103: Variable name "FW_start" doesn't conform to '[a-z_][a-z0-9_]{2,30}$' pattern (invalid-name)
plot_app/vtol_tailsitter.py:67:12: C0103: Variable name "is_FW" doesn't conform to '[a-z_][a-z0-9_]{2,30}$' pattern (invalid-name)
plot_app/vtol_tailsitter.py:78:4: C0103: Variable name "RPY" doesn't conform to '[a-z_][a-z0-9_]{2,30}$' pattern (invalid-name)
```

File: app/plot_app/configured_plots.py (hunk 6 above)
Comment (amandastowers) [reply app/plot_app/configured_plots.py:238]: This is the correction for the rate plots.

File: app/plot_app/vtol_tailsitter.py (hunk 7 above)
Comment (amandastowers) [reply app/plot_app/vtol_tailsitter.py]: made this consistent with other areas where variables are referenced - let me know if you want an explicit flag of some sort set

**Comment by amandastowers** [comment]
OK, think everything above should be corrected now.

File: app/plot_app/configured_plots.py (hunk 6 above)
Comment (bkueng) [reply app/plot_app/configured_plots.py:238]: I was referring only to the `try...except`. I would rather explicitly check for `if rates[axis] is not None:`

**Review by bkueng — COMMENTED** [review]
Almost good now.
Can you resolve the remaining CI errors:
```
************* Module plot_app.configured_plots
plot_app/configured_plots.py:196:9: C0103: Variable name "RPY" doesn't conform to '[a-z_][a-z0-9_]{2,30}$' pattern (invalid-name)
************* Module plot_app.vtol_tailsitter
plot_app/vtol_tailsitter.py:9:69: C0303: Trailing whitespace (trailing-whitespace)
plot_app/vtol_tailsitter.py:12:63: C0303: Trailing whitespace (trailing-whitespace)
plot_app/vtol_tailsitter.py:22:0: C0303: Trailing whitespace (trailing-whitespace)
plot_app/vtol_tailsitter.py:36:0: C0303: Trailing whitespace (trailing-whitespace)
plot_app/vtol_tailsitter.py:43:0: C0303: Trailing whitespace (trailing-whitespace)
plot_app/vtol_tailsitter.py:77:8: C0303: Trailing whitespace (trailing-whitespace)
plot_app/vtol_tailsitter.py:83:0: C0303: Trailing whitespace (trailing-whitespace)
plot_app/vtol_tailsitter.py:110:0: C0303: Trailing whitespace (trailing-whitespace)
plot_app/vtol_tailsitter.py:112:0: C0303: Trailing whitespace (trailing-whitespace)
plot_app/vtol_tailsitter.py:115:0: C0303: Trailing whitespace (trailing-whitespace)
```

**Comment by amandastowers** [comment]
Should be resolved now

**Review by bkueng — APPROVED** [review]
Looks nice now
